GVFS/GVFS.Common/Tracing/TelemetryDaemonEventListener.cs

```diff
@@ -47,14 +46,27 @@ public static TelemetryDaemonEventListener CreateIfEnabled(string gitBinRoot, st
 
         public override void Dispose()
         {
-            this.pipeClient?.Dispose();
+            if (this.pipeWriter != null)
+            {
+                this.pipeWriter.StateChanged -= this.OnPipeWriterStateChanged;
+                this.pipeWriter.Stop();
+                this.pipeWriter.Dispose();
+                this.pipeWriter = null;
+            }
+
             base.Dispose();
         }
 
         protected override void RecordMessageInternal(TraceEventMessage message)
         {
-            string json = this.SerializeMessage(message);
-            this.SendMessage(json);
+            string pipeMessage = this.CreatePipeMessage(message);
+
+            bool dropped = !this.pipeWriter.TryEnqueue(pipeMessage);
+
+            if (dropped)
+            {
+                this.RaiseListenerFailure("Pipe delivery queue is full. Message was dropped.");

```

Comment:
Didn't find great documentation on when TryAdd would return false.  Seems kinda scary we can fill the queue and drop messages.  Is this the instance where we return false?

Reply:
> Didn't find great documentation on when TryAdd would return false.

[`TryAdd(T)`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.blockingcollection-1.tryadd?view=netframework-4.7.2#System_Collections_Concurrent_BlockingCollection_1_TryAdd__0_)

> **Remarks**
> If the collection is a bounded collection, and is full, this method immediately returns false without adding the item.

--

> Seems kinda scary we can fill the queue and drop messages. Is this the instance where we return false?

Yes. In the old implementation we didn't have any throttling since we are synchronously sending (attempting to send) messages over the pipe. Now we're sending messages in the background it is possible we might be getting more messages in faster than can be drained, so we need to prevent an explosion of memory/forever increasing message queue.

This is only used in the telemetry event listener so the events we're happy with dropping - they're not critical. We're already saying dropping a message on send _failure_ is OK.
At least here we'll be surfacing (to the other event listeners) that the message queue is too full.